### PR TITLE
Change configurations api to new programming model

### PR DESCRIPTION
**Pull request by stonezdj**
Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

**Review by heww — APPROVED** [review]

File: src/controller/config/controller.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+			val = string(valByte)
+		}
+		result[item.Name] = &config.Value{
+			Val:      val,
+			Editable: true,
+		}
+	}
+
+	// default value for ScanAllPolicy
+	if _, ok := result[common.ScanAllPolicy]; !ok {
```
Comment (wy65701436) [inline src/controller/config/controller.go:201]: do we still need this? @heww

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/pkg/config/db/dao/dao.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	if n == 0 {
+		return nil, nil
+	}
+	return p, nil
+}
+
+// isDupRecErr checks if the error is due to a duplication of record, currently this
+// works only for pgSQL
+func isDupRecErr(e error) bool {
```
Comment (wy65701436) [inline src/pkg/config/db/dao/dao.go:112]: the lib/orm has the method.

File: src/pkg/config/db/dao/dao.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		if entry.Key == common.LDAPGroupAdminDn {
+			entry.Value = utils.TrimLower(entry.Value)
+		}
+		tempEntry := models.ConfigEntry{}
+		tempEntry.Key = entry.Key
+		tempEntry.Value = entry.Value
+		created, _, err := o.ReadOrCreate(&tempEntry, "k")
+		if err != nil && !isDupRecErr(err) {
+			log.Errorf("Error create configuration entry: %v", err)
+			return err
```
Comment (wy65701436) [inline src/pkg/config/db/dao/dao.go:77]: you can use the lib/err here, err.Wrap()

File: src/server/v2.0/handler/config.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+type configAPI struct {
+	BaseAPI
+	controller config.Controller
+}
+
+func newConfigAPI() *configAPI {
+	return &configAPI{controller: config.Ctl}
+}
+
+func (c configAPI) GetConfigurations(ctx context.Context, params configure.GetConfigurationsParams) middleware.Responder {
```
Comment (wy65701436) [inline src/server/v2.0/handler/config.go:39]: func (c *configAPI)

File: src/server/v2.0/handler/config.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		return c.SendError(ctx, err)
+	}
+	return configure.NewPutConfigurationsOK()
+}
+
+func (c configAPI) GetInternalconfig(ctx context.Context, params configure.GetInternalconfigParams) middleware.Responder {
+	sec, exist := security.FromContext(ctx)
+	if !exist || !sec.IsSolutionUser() {
+		return c.SendError(ctx, errors.UnauthorizedError(nil).WithMessage("only internal service is allowed to call this API"))
+	}
```
Comment (wy65701436) [inline src/server/v2.0/handler/config.go:92]: add a method in base? RequireSolutionUserAccess()?

File: src/server/v2.0/handler/config.go (hunk 4 above)
Comment (stonezdj) [reply src/server/v2.0/handler/config.go:39]: Updated

File: src/pkg/config/db/dao/dao.go (hunk 3 above)
Comment (stonezdj) [reply src/pkg/config/db/dao/dao.go:77]: Updated

File: src/server/v2.0/handler/config.go (hunk 5 above)
Comment (stonezdj) [reply src/server/v2.0/handler/config.go:92]: Updated

File: src/pkg/config/db/dao/dao.go (hunk 2 above)
Comment (stonezdj) [reply src/pkg/config/db/dao/dao.go:112]: Updated

File: src/controller/config/controller.go (hunk 1 above)
Comment (heww) [reply src/controller/config/controller.go:201]: I really don't know whether it's needed for us.

Comment (stonezdj) [reply src/controller/config/controller.go:201]: keep it for now

**Pull request merged**